### ISSUE TITLE
Adding EMA to karras diffusion model

### ISSUE THREAD

**Issue by Joshua-Miller-161** [Joshua-Miller-161/keras_diffusion_cNCSNpp_zarr#9]
@claude it is common to use an exponential moving average to smooth out the training weights. This is some code from a successful model:                             integrate this into the karras diffusion model    DO NOT DIRECTLY COPY ScoreModelLightningModule, this is just an example of how the EMA was implemented in other code  # Modified from https://raw.githubusercontent.com/fadel/pytorch_ema/master/torch_ema/ema.py

from __future__ import division
from __future__ import unicode_literals

import torch

# Partially based on: https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/training/moving_averages.py

class ExponentialMovingAverage:
    def __init__(self, parameters, decay, use_num_updates=True):
        if decay < 0.0 or decay > 1.0:
            raise ValueError('Decay must be between 0 and 1')
        self.decay = decay
        self.num_updates = 0 if use_num_updates else None
        self.shadow_params = [p.clone().detach().to(p.device) for p in parameters if p.requires_grad]
        self.collected_params = []

    def update(self, parameters):
        if self.decay == 1:
            self.shadow_params = [p.clone().detach().to(p.device) for p in parameters if p.requires_grad]
        else:
            decay = self.decay
            if self.num_updates is not None:
                self.num_updates += 1
                decay = min(decay, (1 + self.num_updates) / (10 + self.num_updates))
            one_minus_decay = 1.0 - decay
            with torch.no_grad():
                parameters = [p for p in parameters if p.requires_grad]
                for s_param, param in zip(self.shadow_params, parameters):
                    if s_param.device != param.device:
                        s_param.data = s_param.data.to(param.device)
                    s_param.sub_(one_minus_decay * (s_param - param))

    def copy_to(self, parameters):
        parameters = [p for p in parameters if p.requires_grad]
        for s_param, param in zip(self.shadow_params, parameters):
            if s_param.device != param.device:
                s_param.data = s_param.data.to(param.device)
            param.data.copy_(s_param.data)

    def store(self, parameters):
        self.collected_params = [p.clone().detach().to(p.device) for p in parameters]

    def restore(self, parameters):
        for c_param, param in zip(self.collected_params, parameters):
            if c_param.device != param.device:
                c_param = c_param.to(param.device)
            param.data.copy_(c_param.data)

    def state_dict(self):
        return {
            'decay': self.decay,
            'num_updates': self.num_updates,
            'shadow_params': [p.cpu() for p in self.shadow_params]
        }

    def load_state_dict(self, state_dict, device=None):
        self.decay = state_dict['decay']
        self.num_updates = state_dict['num_updates']
        self.shadow_params = [
            p.to(device) if device is not None else p
            for p in state_dict['shadow_params']
        ]                 import sys
sys.dont_write_bytecode = True
import pytorch_lightning as pl
import torch
import os
import torch.distributed as dist
import logging

logger = logging.getLogger(__name__)

from .ema import ExponentialMovingAverage
from .utils import create_model, is_main_process
from .losses import get_loss
from .sde_lib import get_sde
from .optimizers import get_optimizer
from .lr_schedulers import get_lr_scheduler
#====================================================================
class ScoreModelLightningModule(pl.LightningModule):
    def __init__(self, config):
        super().__init__()
        self.config = config
        
        logger.info(" >> >> INSIDE lightningModule config.deterministic %s, sde: %s", config.deterministic, config.training.sde)

        base_model = create_model(config)
        self.model = torch.compile(base_model, options={"triton.cudagraphs": True}) # PyTorch 2 - comment out if errors n stuff

        #self.model = create_model(config)
        self.sde   = get_sde(config)
        self.ema   = ExponentialMovingAverage(self.model.parameters(),
                                              decay=config.model.ema_rate)
        
        print(" >> >> INSIDE lightningModule", type(self.sde))
        
        self.train_loss_fn = get_loss(self.sde, True, config)
        self.val_loss_fn   = get_loss(self.sde, False, config)
        self.batch_size = config.training.batch_size
        self.val_losses = []
        self.train_losses = []

    def forward(self, x, cond, time_cond):
        return self.model.forward(x, cond, time_cond)

    def on_train_batch_end(self, outputs, batch, batch_idx):
        self.ema.update(self.model.parameters())

    def transfer_batch_to_device(self, batch, device, dataloader_idx=0):
        cond, target, time = batch
        cond   = cond.to(device, non_blocking=True)
        target = target.to(device, non_blocking=True)
        time   = time #.to(device, non_blocking=True)
        return cond, target, time

    def training_step(self, batch, batch_idx):
        rank = dist.get_rank() if dist.is_available() and dist.is_initialized() else 0
        self._batch_counter += 1
        #if (self._batch_counter % 50) == 0:
        #    logger.info(" >> INSIDE lightningModuleEMA training_step [rank %d] processed %d batches so far (batch_idx=%s)", rank, self._batch_counter, batch_idx)
        
        cond, target, time = batch

        train_loss = self.train_loss_fn(self.model, target, cond)
        self.train_losses.append(train_loss.detach())
        #self.log("train_loss", loss, prog_bar=True, logger=True, batch_size=self.batch_size)
        return train_loss

    def on_train_epoch_end(self):
        # Log average training loss
        if self.train_losses:
            avg_train_loss = torch.stack(self.train_losses).mean()
            if self.trainer.global_rank == 0:
                self.log("train_loss", avg_train_loss, prog_bar=True, logger=True, on_epoch=True, on_step=False, rank_zero_only=True)
                print(f" >> >> Epoch {self.current_epoch} - train_loss: {avg_train_loss}")
            self.train_losses.clear()

        # Log the learning rate at the end of each epoch
        optimizer = self.optimizers()
        if isinstance(optimizer, list):
            optimizer = optimizer[0]
        lr = optimizer.param_groups[0]['lr']
        if self.trainer.global_rank == 0:
            self.log("lr", lr, prog_bar=True, logger=True, on_epoch=True, on_step=False, rank_zero_only=True)
            print(f" >> >> Epoch {self.current_epoch} - lr: {lr}")

        rank = dist.get_rank() if dist.is_available() and dist.is_initialized() else 0
        logger.info(" >> INSIDE lightningModuleEMA on_train_epoch_end [rank %d] EPOCH END: processed %d batches this epoch", rank, getattr(self, "_batch_counter", -1))
        
    def validation_step(self, batch, batch_idx):
        #if is_main_process():
            #logger.info("))))))))))))))))))))))))) VALIDATION STEP ((((((((((((((((((((((((((((") 
            #logger.info("))))))))))))))))))))))))) VALIDATION STEP ((((((((((((((((((((((((((((") 
            #logger.info("))))))))))))))))))))))))) VALIDATION STEP ((((((((((((((((((((((((((((") 
            #logger.info("))))))))))))))))))))))))) VALIDATION STEP ((((((((((((((((((((((((((((") 
        
        cond, target, time = batch
        val_loss = self.val_loss_fn(self.model, target, cond)
        self.val_losses.append(val_loss.detach())
        return val_loss

    def on_validation_epoch_end(self):
        # Apply EMA weights
        self.ema.store(self.model.parameters())
        self.ema.copy_to(self.model.parameters())

        # Compute and log average validation loss
        if self.val_losses:
            avg_val_loss = torch.stack(self.val_losses).mean()
            if self.trainer.global_rank == 0:
                self.log("val_loss", avg_val_loss, prog_bar=True, sync_dist=True, logger=True, on_epoch=True, on_step=False, rank_zero_only=True)

        # Restore original weights
        self.ema.restore(self.model.parameters())
        # Clear the buffer
        self.val_losses.clear()

    def on_save_checkpoint(self, checkpoint):
        checkpoint["ema"] = self.ema.state_dict()
        checkpoint["step"] = self.global_step
        checkpoint["epoch"] = self.current_epoch
        #checkpoint["location_params"] = self.config.location_params.state_dict()

    def on_load_checkpoint(self, checkpoint):
        if "ema" in checkpoint:
            self.ema.load_state_dict(checkpoint["ema"], device=self.device)
        if "location_params" in checkpoint:
            self.config.location_params.load_state_dict(checkpoint["location_params"])

    def configure_optimizers(self):
        optimizer = get_optimizer(self.config, self.model.parameters())
        scheduler = get_lr_scheduler(optimizer, self.config)
        return {
            "optimizer": optimizer,
            "lr_scheduler": {
                "scheduler": scheduler,
                "interval": "epoch",
                "frequency": 1,
                "monitor": "val_loss",
            },
        }

    def on_train_epoch_start(self):
        rank = dist.get_rank() if dist.is_available() and dist.is_initialized() else 0
        self._batch_counter = 0
        logger.info(" >> INSIDE lightningModuleEMA on_train_epoch_start [rank %d pid %d] on_train_epoch_start", rank, os.getpid())
                   DO NOT DIRECTLY COPY ScoreModelLightningModule, this is just an example of how the EMA was implemented in other code
